From f7ef6cf4615106062f2ce4dcf1f39668d7fb1959 Mon Sep 17 00:00:00 2001
From: Liu Ying <victor.liu@nxp.com>
Date: Wed, 13 Sep 2017 17:40:18 +0800
Subject: [PATCH 4521/5242] MLK-19413-10 gpu: imx: dpu: Add helpers to peek at
 auxiliary display submodules

commit  601864bc81176c3883d62b8c9d01728f004ca7cc from
https://source.codeaurora.org/external/imx/linux-imx.git

This patch adds dpu_aux_{unit}_peek() helpers so that callers
may peek at auxiliary display submodules.

Signed-off-by: Liu Ying <victor.liu@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/imx/dpu/dpu-constframe.c |   13 +++++++++++++
 drivers/gpu/imx/dpu/dpu-disengcfg.c  |    6 ++++++
 drivers/gpu/imx/dpu/dpu-extdst.c     |   13 +++++++++++++
 drivers/gpu/imx/dpu/dpu-framegen.c   |    6 ++++++
 drivers/gpu/imx/dpu/dpu-tcon.c       |    6 ++++++
 include/video/dpu.h                  |    5 +++++
 6 files changed, 49 insertions(+)

diff --git a/drivers/gpu/imx/dpu/dpu-constframe.c b/drivers/gpu/imx/dpu/dpu-constframe.c
index 57206bb..e25bd7d 100644
--- a/drivers/gpu/imx/dpu/dpu-constframe.c
+++ b/drivers/gpu/imx/dpu/dpu-constframe.c
@@ -171,6 +171,19 @@ void dpu_cf_put(struct dpu_constframe *cf)
 }
 EXPORT_SYMBOL_GPL(dpu_cf_put);
 
+struct dpu_constframe *dpu_aux_cf_peek(struct dpu_constframe *cf)
+{
+	unsigned int aux_id = cf->id ^ 1;
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(cf_ids); i++)
+		if (cf_ids[i] == aux_id)
+			return cf->dpu->cf_priv[i];
+
+	return NULL;
+}
+EXPORT_SYMBOL_GPL(dpu_aux_cf_peek);
+
 void _dpu_cf_init(struct dpu_soc *dpu, unsigned int id)
 {
 	struct dpu_constframe *cf;
diff --git a/drivers/gpu/imx/dpu/dpu-disengcfg.c b/drivers/gpu/imx/dpu/dpu-disengcfg.c
index 76e2188..58b65c2 100644
--- a/drivers/gpu/imx/dpu/dpu-disengcfg.c
+++ b/drivers/gpu/imx/dpu/dpu-disengcfg.c
@@ -115,6 +115,12 @@ void dpu_dec_put(struct dpu_disengcfg *dec)
 }
 EXPORT_SYMBOL_GPL(dpu_dec_put);
 
+struct dpu_disengcfg *dpu_aux_dec_peek(struct dpu_disengcfg *dec)
+{
+	return dec->dpu->dec_priv[dec->id ^ 1];
+}
+EXPORT_SYMBOL_GPL(dpu_aux_dec_peek);
+
 void _dpu_dec_init(struct dpu_soc *dpu, unsigned int id)
 {
 }
diff --git a/drivers/gpu/imx/dpu/dpu-extdst.c b/drivers/gpu/imx/dpu/dpu-extdst.c
index 09a36a9..d69ab5f 100644
--- a/drivers/gpu/imx/dpu/dpu-extdst.c
+++ b/drivers/gpu/imx/dpu/dpu-extdst.c
@@ -440,6 +440,19 @@ void dpu_ed_put(struct dpu_extdst *ed)
 }
 EXPORT_SYMBOL_GPL(dpu_ed_put);
 
+struct dpu_extdst *dpu_aux_ed_peek(struct dpu_extdst *ed)
+{
+	unsigned int aux_id = ed->id ^ 1;
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(ed_ids); i++)
+		if (ed_ids[i] == aux_id)
+			return ed->dpu->ed_priv[i];
+
+	return NULL;
+}
+EXPORT_SYMBOL_GPL(dpu_aux_ed_peek);
+
 void _dpu_ed_init(struct dpu_soc *dpu, unsigned int id)
 {
 	struct dpu_extdst *ed;
diff --git a/drivers/gpu/imx/dpu/dpu-framegen.c b/drivers/gpu/imx/dpu/dpu-framegen.c
index 9f28c13..06c1d2f 100644
--- a/drivers/gpu/imx/dpu/dpu-framegen.c
+++ b/drivers/gpu/imx/dpu/dpu-framegen.c
@@ -551,6 +551,12 @@ void dpu_fg_put(struct dpu_framegen *fg)
 }
 EXPORT_SYMBOL_GPL(dpu_fg_put);
 
+struct dpu_framegen *dpu_aux_fg_peek(struct dpu_framegen *fg)
+{
+	return fg->dpu->fg_priv[fg->id ^ 1];
+}
+EXPORT_SYMBOL_GPL(dpu_aux_fg_peek);
+
 void _dpu_fg_init(struct dpu_soc *dpu, unsigned int id)
 {
 	struct dpu_framegen *fg;
diff --git a/drivers/gpu/imx/dpu/dpu-tcon.c b/drivers/gpu/imx/dpu/dpu-tcon.c
index 0e66a94..4c87cd1 100644
--- a/drivers/gpu/imx/dpu/dpu-tcon.c
+++ b/drivers/gpu/imx/dpu/dpu-tcon.c
@@ -229,6 +229,12 @@ void dpu_tcon_put(struct dpu_tcon *tcon)
 }
 EXPORT_SYMBOL_GPL(dpu_tcon_put);
 
+struct dpu_tcon *dpu_aux_tcon_peek(struct dpu_tcon *tcon)
+{
+	return tcon->dpu->tcon_priv[tcon->id ^ 1];
+}
+EXPORT_SYMBOL_GPL(dpu_aux_tcon_peek);
+
 void _dpu_tcon_init(struct dpu_soc *dpu, unsigned int id)
 {
 }
diff --git a/include/video/dpu.h b/include/video/dpu.h
index 261d59d..bf31e0d 100644
--- a/include/video/dpu.h
+++ b/include/video/dpu.h
@@ -501,12 +501,14 @@ void constframe_constantcolor(struct dpu_constframe *cf, unsigned int r,
 shadow_load_req_t constframe_to_shdldreq_t(struct dpu_constframe *cf);
 struct dpu_constframe *dpu_cf_get(struct dpu_soc *dpu, int id);
 void dpu_cf_put(struct dpu_constframe *cf);
+struct dpu_constframe *dpu_aux_cf_peek(struct dpu_constframe *cf);
 
 /* Display Engine Configuration Unit */
 struct dpu_disengcfg;
 void disengcfg_polarity_ctrl(struct dpu_disengcfg *dec, unsigned int flags);
 struct dpu_disengcfg *dpu_dec_get(struct dpu_soc *dpu, int id);
 void dpu_dec_put(struct dpu_disengcfg *dec);
+struct dpu_disengcfg *dpu_aux_dec_peek(struct dpu_disengcfg *dec);
 
 /* External Destination Unit */
 struct dpu_extdst;
@@ -535,6 +537,7 @@ void constframe_constantcolor(struct dpu_constframe *cf, unsigned int r,
 u32 extdst_perfresult(struct dpu_extdst *ed);
 struct dpu_extdst *dpu_ed_get(struct dpu_soc *dpu, int id);
 void dpu_ed_put(struct dpu_extdst *ed);
+struct dpu_extdst *dpu_aux_ed_peek(struct dpu_extdst *ed);
 
 /* Fetch Decode Unit */
 int fetchdecode_pixengcfg_dynamic_src_sel(struct dpu_fetchunit *fu,
@@ -606,6 +609,7 @@ void framegen_read_timestamp(struct dpu_framegen *fg,
 void framegen_disable_clock(struct dpu_framegen *fg);
 struct dpu_framegen *dpu_fg_get(struct dpu_soc *dpu, int id);
 void dpu_fg_put(struct dpu_framegen *fg);
+struct dpu_framegen *dpu_aux_fg_peek(struct dpu_framegen *fg);
 
 /* Horizontal Scaler Unit */
 struct dpu_hscaler;
@@ -653,6 +657,7 @@ void layerblend_pixengcfg_clken(struct dpu_layerblend *lb,
 void tcon_cfg_videomode(struct dpu_tcon *tcon, struct drm_display_mode *m);
 struct dpu_tcon *dpu_tcon_get(struct dpu_soc *dpu, int id);
 void dpu_tcon_put(struct dpu_tcon *tcon);
+struct dpu_tcon *dpu_aux_tcon_peek(struct dpu_tcon *tcon);
 
 /* Vertical Scaler Unit */
 struct dpu_vscaler;
-- 
1.7.9.5

